docs/architecture/incremental.md

```diff
@@ -0,0 +1,23 @@
+# Configuration: Streams & Incremental Syncs
+
+## AirbyteStream
+* The `AirbyteStream` has two sections:
+    * The `discovered` section describes information and metadata about the data in the source. This section is generated by the source connector. Outside of the source, this section should be treated as read only.
+    * The `configuration` section describes functionality that an end user would like applied on the sync. This section is configured by the end-user, and is never touched by the source.
+
+## Discovered
+
+Currently this section contains two fields: `json_schema` and `metadata`.
+* `json_schema` describes the structure of the data in a source.
+* `metadata` describes other information about the data in the source. For example what features are supported for the stream. Currently the `metadata` object contains two fields.
+    * `supported_sync_modes` which describes which sync modes are supported for a stream.
+    * `default_comparable` which describes what field should be used to determine if a record should be included in a sync. If this is set, the end user does not need to specify their own comparable. If it is not set, then the end user does need to specify it. If it is set and the user sets their own, the user's value overrides the default value.
+
+## Configuration
+* This section currently has two fields: `sync_mode` and `comparable`.
+* `sync_mode` which describes how the data will be synced (e.g. `full_refresh` or `incremental`).
+* `comparable` which is the path to the field that will be used to determine if a record is new or modified since the last sync. This field is REQUIRED if `sync_mode` is `incremental`. Otherwise it is ignored.
```

Comment:
Should this be specified as a `oneOf` to actually encode in the protocol that the comparable field is required for the sync mode?

Reply:
if oneof worked well in java, i'd say yes. we've had so much frustration with oneof and java that i'm really trying to avoid it. we have full control over when these fields are configured and read that we can validate in application code. not ideal, but at least not too risky.

Reply:
talked offline. we can't get java codegen to work with this.